chore: clean up ShellResult usage

addaleax opened the pull request:
Provide more encapsulation by using a fixed interface for other
packages to work with `shell-api`, without the need to use internal
methods of `shell-api` objects.

- Inside `shell-api`:
  - Consistently use `async [asPrintable]()`, `[resultSource]`,
    `[shellApiType]` (all as symbol properties)
    - These are all non-exported, local symbol properties, i.e.
      not for outside access
  - Deprecate `_asPrintable`
  - Deprecate `ShellResult.value` in favor of more expressively named
    `.rawValue` and `.printable` (`.value` is very generic, and in
    particular I would have expected it to be the evaluation result
    and not a formatting-ready version of it)
  - Remove `[asShellResult]()`
- On the API surface of `shell-api`:
  - Remove `asShellResult` and `shellApiType`
  - Provide `getShellApiType()` and `toShellResult()` as the sole APIs
    to turn an evaluation result into a `ShellResult`
  - In particular, this applies for *all* values – no special handling
    is necessary in `shell-evaluator` anymore

This is partly in preparation for MONGOSH-237, so that we do not start
introducing calls to half-internal methods like `_asPrintable()`
in more places than we already do.

mcasimir commented on packages/shell-api/src/decorators.ts:
```diff
@@ -36,27 +36,81 @@ export interface ShellResultSourceInformation {
 export class ShellApiClass implements ShellApiInterface {
   help: any;
-  [asShellResult](): any {
+  get [shellApiType](): string {
```
@aherlihy did we have issues with getters and setters in the past with he java shell?

aherlihy replied:
Exactly, you would have to update the java shell since they use this method directly (which is why it was the only non-symbol of these internal methods)

addaleax replied:
Are you thinking about `_asPrintable()` here? That is still left as an alias for the java shell (`asShellResult` is not mentioned in their source, and always was a symbol)

aherlihy replied:
Yes I was! Sorry the comment was super unclear. I just meant that the reason _asPrintable was different from the others was because of the java shell

mcasimir commented on packages/shell-api/src/decorators.ts:
```diff
@@ -36,27 +36,81 @@ export interface ShellResultSourceInformation {
     return Object.assign({}, this);
   }
+
+  /// @deprecated: Use `toShellResult(value).printable` instead.
```
I'd just drop it and fix it in the consumers, not even sure anyone is using it right now, didn't Jetbrains switch to use the `ShellEvaluator`?

addaleax replied:
They are still calling `_asPrintable()` in a few places right now … I can try to figure out how to move them away from it, but for me, the `java-shell` test suite is not passing locally on `master` to begin with

mcasimir commented on packages/shell-api/src/enums.ts:
```diff
@@ -26,8 +26,11 @@ export enum DBQueryOption {
+export const asPrintable = Symbol.for('@@mongosh.asPrintable');
+export const namespaceInfo = Symbol.for('@@mongosh.namespaceInfo');
+
+/// @deprecated toShellResult() returns both the raw value and the printable version now.
```
same here, would not bother deprecating and just try to make the change in the java shell, then drop a message in mongosh-jetbrains.

mcasimir reviewed (commented):
Looks like a good improvement to me! @aherlihy though is probably a better reviewer here :). Please before merging make sure getter and setters won't cause issues with Jetbrains, i have a vague memory of that being a thing. Also probably we could spare the deprecation if the change in Kotlin is simple enough.

Otherwise let's make a ticket for that and make sure we'd follow up and don't leave it around.

aherlihy commented:
Hi! Have you tried running the java shell tests? In the repo there are instructions for how to run them (it's pretty straight forward)

addaleax commented:
> Hi! Have you tried running the java shell tests? In the repo there are instructions for how to run them (it's pretty straight forward)

Yeah, I did, but they failed for me locally even on `master` … maybe that changed after #372 this morning, the description at least says that they do pass now :)

aherlihy commented:
They do pass for me on current master, you have to create a user before the tests run on the DB you're running locally. So if you start a standalone at 27017 and run `use admin; db.createRole({user: 'admin', pwd: 'admin', roles: ['root']})` then the tests should pass. (It won't be forever - hopefully soon we'll add the java tests to the CI so we can automatically tell if something is off. Which means we'll have to update the java tests so they set up / tear down the servers they test against, but that's future work)

I rebased your branch with master and it looks like the tests are passing, so it seems there's nothing to worry about there :)

addaleax commented:
> I rebased your branch with master (locally) and it looks like the tests are passing, so it seems there's nothing to worry about there :)

Yeah, I’m currently trying to figure out how to get rid of `_asPrintable` there as well, as @mcasimir suggested – seems like that should be doable :)

addaleax commented:
Okay, so:

- Removed `_asPrintable` as a non-symbol/semi-public method altogether, including from the Java shell
- Removed the special handling for the Cursor classes that was introduced in https://github.com/mongodb-js/mongosh/pull/341, the Java shell doesn’t need it anymore

The only remaining deprecation introduced here is the one for `ShellResult.value` here, the VSCode extension uses that, but we can also remove it altogether here and just take care of that at the next mongosh update in VSCode.

aherlihy commented on packages/java-shell/src/main/js/all.js:
```diff
@@ -4,5 +4,7 @@ const ShellEvaluator = require('../../../../shell-evaluator/').default;
     ShellInternalState: ShellApi.ShellInternalState,
-    ShellEvaluator: ShellEvaluator
+    ShellEvaluator: ShellEvaluator,
+    toShellResult: ShellApi.toShellResult,
```
Nice!

aherlihy commented on packages/shell-api/src/enums.ts:
```diff
@@ -26,8 +26,8 @@ export enum DBQueryOption {
-// export const asPrintable = Symbol('asPrintable');
-export const asShellResult = Symbol.for('@@mongosh.asShellResult');
+// TODO: This requires changes to java-shell.
+export const asPrintable = Symbol.for('@@mongosh.asPrintable');
```
I think you can get rid of the TODO here since it's been updated

addaleax replied:
Ah right, thanks for noticing – done! :)

aherlihy commented on packages/shell-api/src/result.ts:
```diff
@@ -102,32 +97,10 @@ export class DeleteResult extends ShellApiClass {
-      },
-      enumerable: false
-    });
+    addHiddenDataProperty(this, shellApiType, 'CursorIterationResult');
```
Much nicer!

aherlihy reviewed (approved):
Looks great! I always thought this was a bit clunky and it's much better now.

kornilova203 commented on packages/java-shell/src/main/kotlin/com/mongodb/mongosh/MongoShellContext.kt:
```diff
@@ -161,13 +165,31 @@ internal class MongoShellContext(client: MongoClient) {
+    }
+
+    fun extract(printable: Value?, rawValue: Value? = null): MongoShellResult<*> {
+        var v: Value
+        var type: String? = null
+        if (printable == null) {
+            type = getShellApiType(rawValue as Value)
+            val shellResult = toShellResult(rawValue as Value)
+            v = shellResult["printable"]!!
+        } else {
+            v = printable
+        }
```
We can use `val` here
```suggestion
        val v: Value
        val type: String?
        if (printable == null) {
            type = getShellApiType(rawValue as Value)
            v = toShellResult(rawValue)["printable"]!!
        } else {
            type = null
            v = printable
        }
```

kornilova203 reviewed (changes requested):
Overall looks good
I'm happy that shell implementations now have more control over results handling

addaleax commented:
@kornilova203 Thanks for the review, made the changes you suggested :)
